packages/nuxt/src/module.ts

```diff
@@ -1,70 +1,62 @@
 /**
  * @module @pinia/nuxt
  */
-// import { isVue2 } from 'vue-demi'
-import type { Pinia } from 'pinia'
-import type { Context, Module } from '@nuxt/types'
-
-export interface PiniaNuxtOptions {
+import {
+  defineNuxtModule,
+  addPlugin,
+  isNuxt2,
+  addAutoImport,
+  createResolver,
+} from '@nuxt/kit'
+
+export interface ModuleOptions {
   /**
    * Pinia disables Vuex by default, set this option to `false` to avoid it and
-   * use Pinia alongside Vuex.
+   * use Pinia alongside Vuex (Nuxt 2 only)
    *
    * @default `true`
    */
   disableVuex?: boolean
 }
 
-const DEFAULTS = {
-  disableVuex: true,
-}
-
-export default <Module>function (_options) {
-  const nuxt = this.nuxt
-  const options = {
-    ...DEFAULTS,
-    ...(_options || {}),
-    ...(nuxt.options.pinia || {}),
-  }
-
-  // Disable default Vuex store (options.features only exists in Nuxt v2.10+)
-  if (nuxt.options.features && options.disableVuex) {
-    nuxt.options.features.store = false
-  }
-
-  // make sure we use the mjs for pinia so node doesn't complain about using a module js with an extension that is js
-  // but doesn't have the type: module in its packages.json file
-  nuxt.options.alias.pinia = 'pinia/dist/pinia.mjs'
-
-  this.addPlugin({ src: require.resolve('./plugin.mjs') })
-
-  // transpile pinia for nuxt 2 and nuxt bridge
-  // if (isVue2 && !nuxt.options.build.transpile.includes('pinia')) {
-  //   nuxt.options.build.transpile.push('pinia')
-  // }
-}
-
-declare module '@nuxt/types' {
-  export interface Context {
-    /**
-     * Pinia instance attached to the app.
-     *
-     * @deprecated: use context.$pinia instead
-     */
-    pinia: Pinia
-
-    /**
-     * Pinia instance attached to the app.
-     */
-    $pinia: Pinia
-  }
-}
-
-declare module 'pinia' {
-  export interface PiniaCustomProperties {
-    /**
-     * Nuxt context.
-     */
-    $nuxt: Context
-  }
-}
+export default defineNuxtModule<ModuleOptions>({
+  meta: {
+    name: 'pinia',
+    configKey: 'pinia',
+    compatibility: {
+      nuxt: '^2.0.0 || ^3.0.0',
+      bridge: true,
+    },
+  },
+  defaults: {
+    disableVuex: true,
+  },
+  setup(options, nuxt) {
+    const resolver = createResolver(import.meta.url)
+
+    // Disable default Vuex store (Nuxt v2.10+ only)
+    if (nuxt.options.features && options.disableVuex && isNuxt2()) {
+      nuxt.options.features.store = false
+    }
+
+    // Transpile runtime
+    nuxt.options.build.transpile.push(resolver.resolve('./runtime'))
+
+    // Make sure we use the mjs build for pinia
+    nuxt.options.alias.pinia = 'pinia/dist/pinia.mjs'
+
+    // Add runtime plugin
+    if (isNuxt2()) {
+      addPlugin(resolver.resolve('./runtime/plugin.vue2'))
+    } else {
+      addPlugin(resolver.resolve('./runtime/plugin.vue3'))
+    }
```

Comment:
> Note: By default plugin is prepended to the plugins array. You can use second argument to append (push) instead.

This means, that by default, if we precise an array of modules like so:
```
modules: ['@pinia/nuxt', 'my-module']
```

We'll receive `getActivePinia was called with no active Pinia` error, as any plugin added by `my-module` with default options is going to be executed **before** pinia plugin can call `setActivePinia()`.

See reproduction: https://stackblitz.com/edit/github-ba6rwp?file=nuxt.config.ts

Reply:
You shouldn't depend on module order (please see my comment: https://github.com/nuxt/framework/issues/5835#issuecomment-1181679108). We might improve with a new hook to allow extending pinia via other module plugins.